src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -310,248 +503,354 @@ public void afterSave() {
-        BitbucketApi bitbucket = head instanceof PullRequestSCMHead
-                ? buildBitbucketClient((PullRequestSCMHead) head)
-                : buildBitbucketClient();
-        String branchName = head instanceof PullRequestSCMHead ? ((PullRequestSCMHead) head).getBranchName() : head.getName();
-        List<? extends BitbucketBranch> branches = bitbucket.getBranches();
+    private String findRawNode(String branchName, List<? extends BitbucketBranch> branches, TaskListener listener) {
         for (BitbucketBranch b : branches) {
             if (branchName.equals(b.getName())) {
-                if (b.getRawNode() == null) {
-                    if (getBitbucketServerUrl() == null) {
-                        listener.getLogger().format("Cannot resolve the hash of the revision in branch %s", b.getName());
+                String revision = b.getRawNode();
+                if (revision == null) {
+                    if (BitbucketCloudEndpoint.SERVER_URL.equals(getServerUrl())) {
+                        listener.getLogger().format("Cannot resolve the hash of the revision in branch %s%n",
+                                branchName);
                     } else {
-                        listener.getLogger().format("Cannot resolve the hash of the revision in branch %s. Perhaps you are using Bitbucket Server previous to 4.x", b.getName());
+                        listener.getLogger().format("Cannot resolve the hash of the revision in branch %s. "
+                                        + "Perhaps you are using Bitbucket Server previous to 4.x%n",
+                                branchName);
                     }
                     return null;
                 }
-                if (getRepositoryType() == BitbucketRepositoryType.MERCURIAL) {
-                    return new MercurialRevision(head, b.getRawNode());
-                } else {
-                    return new AbstractGitSCMSource.SCMRevisionImpl(head, b.getRawNode());
-                }
+                return revision;
             }
         }
-        LOGGER.log(Level.WARNING, "No branch found in {0}/{1} with name [{2}]", head instanceof PullRequestSCMHead
-                ? new Object[]{
-                ((PullRequestSCMHead) head).getRepoOwner(),
-                ((PullRequestSCMHead) head).getRepository(),
-                ((PullRequestSCMHead) head).getBranchName()}
-                : new Object[]{repoOwner, repository, head.getName()});
+        listener.getLogger().format("Cannot find the branch %s%n", branchName);
         return null;
     }
 
     @Override
     public SCM build(SCMHead head, SCMRevision revision) {
-        BitbucketRepositoryType repositoryType;
+        BitbucketRepositoryType type;
         if (head instanceof PullRequestSCMHead) {
-            repositoryType = ((PullRequestSCMHead) head).getRepositoryType();
+            type = ((PullRequestSCMHead) head).getRepositoryType();
```

Comment:
because `getRepositoryType()` is only really needed for Bitbucket cloud. I could add a common interface and just check that... but that risks people making assumptions about interfaces and makes it harder to refactor without breaking backwards compatibility